Document about stale resource cleanup (#910)
ceph/ceph-csi: docs/resourse-cleanup.md

# Stale Resource Cleanup

> If the PVC is created with storage class which is having the reclaimPolicy
> as Retain will not delete the PV object and backend omap metadata and backend image.
> Manual deletion of PV will result in stale omap keys, values and also stale cephfs and rbd images.
> You need to cleanup metadata and image separately.

Reviewer: ```suggestion
If the PVC is created with storage class which is having the `reclaimPolicy``
as `Retain` will not delete the PV object ,backend omap metadata and backend image.
Manual deletion of PV will result in stale omap keys, values , cephfs subvolume and rbd image.
You need to cleanup metadata and image separately.
```

### RBD

**Find related image of pvc**:

```[$]kubectl get pvc pvc_name -n namespace -oyaml``` to get pv_name <br>
```[$]kubectl get pv pv_name -oyaml``` to get part of rbd_image_name(the last part of spec.csi.volumeHandle, xxx for example) <br>
```[$]rbd ls -p pool_name | grep xxx``` to get full rbd_image_name

> **Delete metadata and image**:

Reviewer: need to provide an option to delete omap data from rados
steps would be
* Get volume ID from PV
* Get omap keys
* Delete omap Keys
* Delete rbd image /Delete cephfs image
* Delete PV

providing an example would be a good one

Author: Does 'get omap keys' mean using 'block_name_prefix' of rbd_info to get all objects , then remove omapkeys of object one by one ?

Reviewer: https://github.com/ceph/ceph-csi/blob/master/troubleshooting/tools/tracevol.py will help you to understand more on this

Author: I update my patch and thanks for your review.

```[$]rbd image-meta list rbd_image_name -p pool_name``` to get all metadata <br>
```[$]rbd image-meta remove rdb_image_name metadata_key -p pool_name``` to remove metadata <br>

> ```[$]rbd remove rbd_image_name -p pool_name``` to remove image

Reviewer: removing rbd image will remove the image-meta no need to delete it

> ### CephFS

Reviewer: RBD is bold above need to follow same for both

# Stale Resource Cleanup

If the PVC is created with storage class which is having the `reclaimPolicy``
as `Retain` will not delete the PV object , backend omap metadata and backend image.
Manual deletion of PV will result in stale omap keys, values , cephfs subvolume and rbd image.

> You need to cleanup metadata and image separately.

Reviewer 2: Instead of 'you': 'It is required to cleanup ...'

Author: done.

## Steps

### 1. Get volume ID from PV

> * '[$]kubectl get pvc pvc_name -n namespace -owide' to get pv_name <br>

Reviewer 2: Instead of the ' quote, use a single backtick `.
No need to add `<br>`, use markdown style completely. The following lines should be indented by (2 or 4?) spaces (including the triple backticks). You can use https://hackmd.io for testing the formatting.

Author: done

```
$ kubectl get pvc mysql-pvc -owide -n prometheus
NAME        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE   VOLUMEMODE
mysql-pvc   Bound    pvc-bc537af8-67fc-4963-99c4-f40b3401686a   20Gi       RWO            csi-rbd        14d   Filesystem
```

### 2. Get omap keys

* '[$]rados listomapkeys csi.volumes.default -p pool_name | grep pv_name' to get omapkey1 <br>
```

> $ rados listomapkeys csi.volumes.default -p kube_csi | grep pvc-bc537af8-67fc-4963-99c4-f40b3401686a

Reviewer 3: `csi.volumes.default` need not be named with the suffix `default` always. Request a note here to state that the suffix should be the value used for the CLI option [`--instanceid`](https://github.com/ceph/ceph-csi/blob/master/docs/deploy-rbd.md#configuration) in the provisioner (or, nodeplugin daemonset) [deployment](https://github.com/ceph/ceph-csi/blob/2cc59ca41155b21cfcf30999a17e4b029fc2f26c/deploy/rbd/kubernetes/csi-rbdplugin-provisioner.yaml#L105-L111).

Author: done

csi.volume.pvc-bc537af8-67fc-4963-99c4-f40b3401686a
```
* '[$]rados listomapkeys omapkey1 -p pool_name' to get omapkey2 <br>
```

> $ rados listomapkeys csi.volume.dd2473d0-6a8c-11ea-9113-0ad59d995ce7 -p kube_csi

Reviewer 3: The document does not talk about how the UUID of the volume OMap is arrived at `dd2473d0-6a8c-11ea-9113-0ad59d995ce7`. This is the value of the OMap key above, and would help if mentioned.

Author: done

csi.volname
```

### 3. Delete omap keys

* '[$]rados rmomapkey omapkey1 omapkey2 -p pool_name' <br>
```

> $ rados rmomapkey csi.volume.dd2473d0-6a8c-11ea-9113-0ad59d995ce7 csi.volname  -p kube_csi

Reviewer 3: You should remove the entire OMap named csi.volume.<UUID>, instead of only the key within it. I would suggest replacing this with, `rados rm csi.volume.dd2473d0-6a8c-11ea-9113-0ad59d995ce7`

Author: done.

```
* '[$]rados rmomapkey csi.volumes.default omapkey1 -p pool_name' <br>
```
$ rados rmomapkey csi.volumes.default csi.volume.dd2473d0-6a8c-11ea-9113-0ad59d995ce7 -p kube_csi
```

> ### 4. Delete rbd image /Delete cephfs image

Reviewer 2: Maybe: Delete the RBD image or CephFS subvolume

Author: done.

> ### 4. Delete rbd image /Delete cephfs image

Reviewer 3: Swap step 3 and step 4. The linking is from the `csi.volumes.default` -> `csi.volume.<UUID>` -> `image`, hence deletion should be the reverse of the same, making it safer to determine the image again on any errors in the steps.

Author: done
